clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -710,8 +710,11 @@ private boolean appendsInProgress() {
      */
     public void awaitFlushCompletion() throws InterruptedException {
         try {
-            for (ProducerBatch batch : this.incomplete.copyAll())
-                batch.produceFuture.await();
+            // Make a copy of of the request results at the time the flush is called.
+            // We avoid making a copy of the full incomplete batch collections to allow
+            // garbage collection.
```

Comment:
Hmm, we still have a reference to `incomplete` in the class, right? Are we expecting the GC to free that while blocked in this method?

Reply:
We do have a reference to incomplete, however the sender will remove the producer batches from the original incomplete collection. It's not currently doing so because we're making a full copy of incomplete.

Reply:
That makes sense. It's worth clarifying that here.

Reply:
I've tried to improve the error message. LMK what you think.